rfcs/015 - Reworked Compiler Front End.md

```diff
@@ -0,0 +1,179 @@
+```ln
+fn toString(val) {
+  if val.isSome() {
+    return val.getMaybe().toString();
+  } else {
+    return 'none';
+  }
+}
+```
+
+In order to share more nuanced details about the new front-end, here are the steps the compiler might take upon having to use this function:
+1. Assign the variable `val` to type `T`. There are no constraints on the type `T`.
+2. Desugar the if/else. How this is done is unimportant, but the body of the `toString` function will effectively be desugared into (ignore the invalid `conditionTailFn` function body - it's required for `evalcond` to work but its contents won't be checked by the type checker, since this desugaring will actually happen just before generating the Microstatements):
+```ln
+const conditionalTable = newarr(2);
+const condition1Boolean = val.isSome();
+const condition1ThenFn = fn(): U { return val.getMaybe().toString(); };
+condfn(conditionalTable, condition1Boolean, condition1ThenFn);
+const condition2Boolean = true;
+const condition2ThenFn = fn(): U { return 'none'; };
+condfn(conditionalTable, condition2Boolean, condition2ThenFn);
+const conditionTailFn = fn(): U {};
+const conditionResult: Maybe<U> = evalcond(conditionalTable, conditionTailFn);
+const conditionResultUnwrapped: U = conditionResult.getMaybe();
+return conditionResultUnwrapped;
+```
+3. Add the constraint that `val` has type `A` where there's a function called `isSome` with the signature `(A) -> bool`.
+4. Add the constraint that `val` has type `B` where there's a function called `getMaybe` with the signature `(B) -> C`.
+5. Add the constraint that type `C` from `getMaybe` has a function called `toString` with the signature `(C) -> U`.
+6. Add the constraint that type `U` is the same as the concrete type `string` (this is from the `return 'none'` statement).
+7. Reduce type `A` to only have the possibility of being type `Maybe<S>`.
+8. Reduce type `B` to only have the possibility of being type `Maybe<R>`.
+9. Since `Stringifiable` is an already-existing interface that perfectly matches the constraints on type `C`, replace all instances of type `C` with `Stringifiable`.
+10. Since `A` and `B` both are in reference to the same variable, add the constraint that they satisfy each other.
+11. Reduce type `Maybe<R>` to be equivalent to the type `Maybe<Stringifiable>`
+12. `Maybe<Stringifiable>` is the minimal interface value for `val`, so the signature of the `toString` fn is now rewritten as `(Maybe<Stringifiable>) -> U`.
+13. `string` is the concrete type representation for `U`, so the signature of the `toString` fn is now rewritten as `(Maybe<Stringifiable>) -> string`.
+
+(side note: the `getMaybe()` call will also eventually be syntactic sugar, according to [RFC 12][RFC-12]).
+
+### Alternatives Considered
+
+One alternative is to keep the current type system and compiler front-end and adding changes as-needed.
```

Comment:
There are a lot of parts of the current compiler that may still be worth keeping around, though. How module resolution and import/export works is quite a bit of logic to recreate.

I would also argue that the RFC as-is is incomplete. There's a brief statement about opting into the new frontend by naming the file `.lnn` but how does it fit in with the rest of the compiler, how does it do AST parsing, the above-mentioned module resolution, represent scoping, what the intermediate types are (equivalents to Interface/Type/UserFunction/etc) and then whether or not the microstatement generation that is then serialized to AMM is performed or if the Handlers convert themselves to AMM which triggers the UserFunction within to do so which triggers the Statement within, etc, and then if/how constant hoisting and deduplication is handled.

All of that is *a lot* that the current first stage of the compiler does, and throwing it all away for just the type system is a hard sell.

I think another option to consider is "forking" the current `lntoamm` directory by copying it into an `lnntoamm` directory, ripping out the current type system and replacing it.

Reply:
Gotcha - will add that it will be effectively a fork of the current `lntoamm` directory with parsing and module resolution kept intact, but with a new Microstatement generation system that follows this new type system.

Reply:
The fork of the current `lntoamm` makes a lot of sense